x/ibctransfer/whitelisting.go

```diff
@@ -0,0 +1,95 @@
+package ibctransfer
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	transfer "github.com/cosmos/cosmos-sdk/x/ibc/applications/transfer"
+	transfertypes "github.com/cosmos/cosmos-sdk/x/ibc/applications/transfer/types"
+	channeltypes "github.com/cosmos/cosmos-sdk/x/ibc/core/04-channel/types"
+)
+
+func OnRecvPacketWhiteListed(
+	ctx sdk.Context,
+	sdkAppModule transfer.AppModule,
+	packet channeltypes.Packet,
+) (*sdk.Result, []byte, error) {
+
+	var data transfertypes.FungibleTokenPacketData
+	if err := transfertypes.ModuleCdc.UnmarshalJSON(packet.GetData(), &data); err != nil {
+		return nil, nil, sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest, "cannot unmarshal ICS-20 transfer packet data: %s", err.Error())
+	}
+
+	denom := GetMintedDenomFromPacket(packet, data)
+	isWhitelisted := IsWhitelisted(ctx, denom)
+
+	isReturning := IsRecvPacketReturning(packet, data)
+	if !(isReturning || isWhitelisted) {
+		return nil, nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidCoins, "denom not on whitelist")
+	}
+
+	return sdkAppModule.OnRecvPacket(ctx, packet)
+}
+
+func GetMintedDenomFromPacket(packet channeltypes.Packet, data transfertypes.FungibleTokenPacketData) string {
+	// Note: Code and comments taken from SDK transfer keeper,
+	// used here only to determine the token that will be minted.
+
+	if transfertypes.ReceiverChainIsSource(packet.GetSourcePort(), packet.GetSourceChannel(), data.Denom) {
+		// sender chain is not the source, unescrow tokens
+
+		// remove prefix added by sender chain
+		voucherPrefix := transfertypes.GetDenomPrefix(packet.GetSourcePort(), packet.GetSourceChannel())
+		unprefixedDenom := data.Denom[len(voucherPrefix):]
+
+		// coin denomination used in sending from the escrow address
+		denom := unprefixedDenom
+
+		// The denomination used to send the coins is either the native denom or the hash of the path
+		// if the denomination is not native.
+		denomTrace := transfertypes.ParseDenomTrace(unprefixedDenom)
+		if denomTrace.Path != "" {
+			denom = denomTrace.IBCDenom()
+		}
+
+		return denom
+	}
```

Comment:
Cant we just return denom,bool . 
transfertypes.ReceiverChainIsSource is true "",true
transfertypes.ReceiverChainIsSource is false denom,false

And check for whitelist only in the second case .

Reply:
Assuming that we can allow anything that went from sifchain to come back

Reply:
Oh are we planning to keep non 18 decimal denoms in the whitelist too ? And use the same whitelist to get their units ?